Add ArmorTypes to HitShape and adapt damage warheads

reaperrr opened the pull request:
Supersedes #13675.

This allows to assign `Armor` traits to specific `HitShape`s, allowing to make certain parts of actors more vulnerable or durable than others.
Can also be used to simulate directional armor in a good-enough way that this closes #9268 and therefore closes #7718.
Also closes #13755.

Testcase can be found here: https://github.com/reaperrr/OpenRA/tree/hitshape-armor-v2-test

The main problems of #13675 that brought it down, 

> The core of the problem is that this introduces a second channel for applying damage (DoImpact with and without a hitshape) right in the middle of a chain of inheritance and overrides. This makes it very difficult to know which will be used, and means that the same weapon/warheads/vs fired by different means will end up doing different amounts of damage to the same actor.
This feature really needs to be integrated from the bottom up, and not just bolted on the side. **Require all damaging warheads to specify the hitshape**, and **fix the higher level code to be able to choose the appropriate one based on whatever rules make sense in each case**.

should be adressed by this updated version.

All damage warheads now absolutely require an enabled hitshape to deal damage to actors, and now that we have the Polygon shape type, there's no real excuse for an actor to have overlapping shapes anymore, so just taking the shape closest to impact for damage calculations should be a sufficient solution.

reaperrr committed:
Fix FindActorsOnCircle summary

The summary was not entirely correct:
Since FAOC simply adds the largest existing OuterRadius to the specified circle range, it's still possible (very likely, in fact) for this helper to return actors whose HitShape is entirely outside the specified range (*without* the added largest OuterRadius).

GraionDilach commented:
A lint rule to detect overlapping hitshapes would be nice.

pchote commented on OpenRA.Mods.Common/Warheads/DamageWarhead.cs:
```diff
@@ -35,32 +35,48 @@ public override bool IsValidAgainst(Actor victim, Actor firedBy)
-		public virtual void DoImpact(Actor victim, Actor firedBy, IEnumerable<int> damageModifiers)
+		public virtual void DoImpact(Actor victim, Actor firedBy, HitShapeInfo shapeInfo, IEnumerable<int> damageModifiers)
 		{
-			if (!IsValidAgainst(victim, firedBy))
```
Removing this from here doesn't sit well with me - this is public, so anything can call it.

Considering that this is actually effectively only called from three places (`HealthPercentageWarhead` overrides both this method and the thing that calls it) my preference is to remove this completely and inline the `InflictDamage` calls.

pchote commented on OpenRA.Mods.Common/Warheads/HealthPercentageDamageWarhead.cs:
```diff
@@ -30,25 +30,38 @@ public override void DoImpact(WPos pos, Actor firedBy, IEnumerable<int> damageMo
-				DoImpact(victim, firedBy, damageModifiers);
+			{
+				// Cannot be damaged without a Health trait
+				if (!victim.Info.HasTraitInfo<HealthInfo>())
```
Can we move this inside `IsValidAgainst`? Same comment for the other warheads.

pchote commented on OpenRA.Mods.Common/Warheads/HealthPercentageDamageWarhead.cs:
```diff
@@ -30,25 +30,38 @@ public override void DoImpact(WPos pos, Actor firedBy, IEnumerable<int> damageMo
+				if (!IsValidAgainst(victim, firedBy))
+					continue;
+
+				var impactToVictimDistance = new WDist((victim.CenterPosition - pos).Length);
```
You take the `.Length`, wrap it in a `WDist`, only to call `.Length` on that again below!
IMO use `(victim.CenterPosition - pos).Length` directly in the lerp.

reaperrr commented on OpenRA.Mods.Common/Warheads/HealthPercentageDamageWarhead.cs:
```diff
@@ -24,32 +24,40 @@ public class HealthPercentageDamageWarhead : DamageWarhead
+			// FindActors*On*Circle is needed to include actors who are close enough to Spread[0] that their HitShape might intersect with it.
+			// However, we only want to exclude actors where CenterPosition is inside Range[1], so keeping *InCircle is correct there.
+			// TODO: Either properly account for hitshape intersection with Spread[1] instead, or drop ring support entirely.
+			// None of our shipping mods currently use this at time of writing.
```
Note: If neither the TODO nor then 2nd option are good enough for this PR, I'll need help coming up with a way to check the hitshape intersection, as the check needs to work in reverse of what we normally do (here we need to check if the entire shape is *within* a radius), and I can't get my head around what the math would need to look like.

pchote replied:
Lets remove the ring code.  I don't understand why we had this in the first place.

reaperrr commented:
Rebased and updated.

Removed ring code, changed ArmorTypes to BitSet.

abcdefg30 reviewed (dismissed):
Looks good to me. :+1:

pchote commented on OpenRA.Mods.Common/Warheads/HealthPercentageDamageWarhead.cs:
```diff
@@ -18,38 +18,40 @@ namespace OpenRA.Mods.Common.Warheads
+				if (!IsValidAgainst(victim, firedBy))
+					continue;
+
+				var closestSpreadPos = WPos.Lerp(pos, victim.CenterPosition, Spread.Length, (victim.CenterPosition - pos).Length);
```
If `Spread.Length` is bigger than `(victim.CenterPosition - pos).Length` this will lerp past the center of the actor and out the other side!

I'll push a fix to this alongside a few other cleanups.

pchote commented on OpenRA.Mods.Common/Warheads/TargetDamageWarhead.cs:
```diff
@@ -10,49 +10,53 @@
+					continue;
+
+				// Fully damage actors whose HitShape intersects with warhead Spread, otherwise don't damage at all
+				var distance = closestActiveShape.Info.Type.DistanceFromEdge(pos, victim);
```
This doesn't account for the spread at all!

I'll push a fix to this alongside a few other cleanups.

reaperrr committed:
Make HitShape mandatory for damaging actors and refactor warheads.

* Adds support for linking Armor traits to HitShapes.
* Adds spread support to TargetDamageWarhead
* Removes ring-damage support from HealthPercentageDamage
* Removes IsValidAgainst check from DoImpact(Actor victim...) overload
  and instead lets warheads perform the check beforehand
  (to avoid HitShape look-ups on invalid targets).
* Reduces duplication and improves readability of Warhead implementations

pchote commented:
I have updated this on @reaperrr's behalf:
* Removed duplicated `DistanceFromEdge` calls (@abcdefg30's point above)
* Fixed spread calculations on `TargetDamageWarhead` and `HealthPercentageDamageWarhead`
* Made `HealthPercentageDamageWarhead` a subclass of `TargetDamageWarhead` to further reduce duplication
* Squashed commits to avoid broken compilation (which is a real pain if hit during a bisect)
* Fixed a couple of minor code nits (mainly removing redundant variable definitions)

I've tested that regular weapons, MAD Tanks, and MAD Tanks with warheads replaced by `TargetDamageWarhead` appear to work as expected.  There are some big changes in this push, so I would like if @abcdefg30 or @reaperrr could sign-off on this before merging.

pchote commented:
I think this now also closes #13755?

reaperrr commented:
> I think this now also closes #13755?

Yep, updated original post. I'll give this a last look tomorrow.

reaperrr commented:
Changes look good to me and didn't spot any regressions, so lgtm.

reaperrr commented:
[changelog](https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/c1b0202bf8cd8e8ae1e17b5e608200189c68ec3a...41e8cab61c71a643feba5a9176466c077ebeb70e)
